# Assignment 1: Data Parsing, Cleansing and Integration
## Task 3
#### Student Name: Ziqing Yan
#### Student ID: s3749857

Date: 26 August 2021

Version: 1.0

Environment: Python 3 and Jupyter notebook

Libraries used: please include the main libraries you used in your assignment, e.g.,:
* pandas
* numpy
* random


## Introduction
These two data sets have a lot of conflicts in naming, values and duplications. For naming conflicts, I rename them to make sure the names in the two data sets are consistent. For value conflicts, I replace them with proper values. For duplication conflicts, I delete those duplicated. The detail of how I handle them is showed below.

##  Import libraries 

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import random

### 1. Examining and loading data


In [2]:
# Load the data
df1 = pd.read_csv('S3749857_dataset1_solution.csv')
df2 = pd.read_csv('S3749857_dataset2.csv')
# Check the structure of the dataset 2
df2

Job Title  \
0                   Nursing Home Manager  Thames Ditton   
1         Deputy Manager  Nursing Home  RGN  Shrewsbury   
2            Deputy Manager  Nursing Home (RGN)  London   
3           Deputy Manager (RGN)  Nursing Home  Bristol   
4                                     RGN or RMN  Wigan   
...                                                 ...   
4995  Pharmaceutical  Analytical Chemist Jobs in Nor...   
4996     Medical Sales Representative  South East Essex   
4997  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
4998                     Curriculum Leader  Mathematics   
4999                TEACHER OF BUSINESS STUDIES AND LAW   

                        Organisation  Monthly Payment        Contract Type  \
0               Liquid Personnel Ltd          2750.00                  NaN   
1               Liquid Personnel Ltd          2583.33                  NaN   
2               Liquid Personnel Ltd          3000.00                  NaN   
3               Liquid Personnel Ltd          2791.67                  NaN   
4                         Eden Brown          2240.00                  NaN   
...                              ...              ...                  ...   
4995                    Flame Pharma          1916.67            Permanent   
4996  The Vacancy Management Company          1875.00  Fixed Term Contract   
4997                Live Recruitment          2708.33            Permanent   
4998                             NaN           417.17            Permanent   
4999                             NaN          1900.00  Fixed Term Contract   

                           Category              Closing          Location  \
0            Healthcare and Nursing  2013-02-22 15:00:00     Thames Ditton   
1            Healthcare and Nursing  2012-11-14 15:00:00        Shrewsbury   
2            Healthcare and Nursing  2012-12-08 00:00:00            London   
3            Healthcare and Nursing  2013-01-13 15:00:00           Bristol   
4            Healthcare and Nursing  2013-01-24 15:00:00             Wigan   
...                             ...                  ...               ...   
4995         Healthcare and Nursing  2013-10-24 15:00:00  Northern Ireland   
4996         Healthcare and Nursing  2012-06-20 12:00:00                UK   
4997  PR, Advertising and Marketing  2013-12-17 15:00:00     Hertfordshire   
4998                       Teaching  2012-03-15 12:00:00           Wrexham   
4999                       Teaching  2013-05-11 15:00:00         Salisbury   

     Full-Time Equivalent (FTE)              Opening  
0                     FULL TIME  2012-12-24 15:00:00  
1                     FULL TIME  2012-10-15 15:00:00  
2                     FULL TIME  2012-11-08 00:00:00  
3                     FULL TIME  2012-11-14 15:00:00  
4                     FULL TIME  2012-11-25 15:00:00  
...                         ...                  ...  
4995                        NaN  2013-09-24 15:00:00  
4996                        NaN  2012-05-21 12:00:00  
4997                  FULL TIME  2013-11-17 15:00:00  
4998                        NaN  2012-01-15 12:00:00  
4999                        NaN  2013-04-27 15:00:00  

[5000 rows x 9 columns]

Then, check if there is any missing value in each column of df2.

In [3]:
df2.count()

Job Title                     5000
Organisation                  4937
Monthly Payment               5000
Contract Type                 4312
Category                      5000
Closing                       5000
Location                      5000
Full-Time Equivalent (FTE)     643
Opening                       5000
dtype: int64

There are missing values in the `Organisation`, `Contract Type` and `Full-Time Equivalent (FTE)` column. We need to solve this problem in the next steps.

### 2. Resolving schema conflicts

#### Conflict 1: missing conflicts in the Id column (e.g. naming conflicts in xx column and xx column)

In [4]:
# Check the columns in df1
print(df1.columns)
# Check the columns in df2
print(df2.columns)

Index(['Id', 'Title', 'Location', 'Company', 'ContractType', 'ContractTime',
       'Category', 'Salary', 'OpenDate', 'CloseDate', 'SourceName'],
      dtype='object')
Index(['Job Title', 'Organisation', 'Monthly Payment', 'Contract Type',
       'Category', 'Closing', 'Location', 'Full-Time Equivalent (FTE)',
       'Opening'],
      dtype='object')


There is no such column in df2 which correspond the ID column in df1, so we need to create a new column called ID in df2 and assign 8-digit unique values to the new column. (The values in the new column is also unique in df1)

In [5]:
# Generate unique Id for df2
Id = [] # For storing Id for df2

for var in range(0, df2.shape[0]):
    check = True
    while check:
        ran = random.randint(10000000,99999999)
        # If the randomly produced number is not in the Id column in df1 and df2, pick it as the Id value
        if (ran not in list(df1['Id'])) & (ran not in Id):
            Id.append(ran)
            check = False
            
df2.insert(0, 'Id', Id)
df2

Id                                          Job Title  \
0     95056457                Nursing Home Manager  Thames Ditton   
1     54116178      Deputy Manager  Nursing Home  RGN  Shrewsbury   
2     89491746         Deputy Manager  Nursing Home (RGN)  London   
3     31629963        Deputy Manager (RGN)  Nursing Home  Bristol   
4     77606265                                  RGN or RMN  Wigan   
...        ...                                                ...   
4995  63556254  Pharmaceutical  Analytical Chemist Jobs in Nor...   
4996  80341553     Medical Sales Representative  South East Essex   
4997  32782075  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
4998  64933538                     Curriculum Leader  Mathematics   
4999  83618722                TEACHER OF BUSINESS STUDIES AND LAW   

                        Organisation  Monthly Payment        Contract Type  \
0               Liquid Personnel Ltd          2750.00                  NaN   
1               Liquid Personnel Ltd          2583.33                  NaN   
2               Liquid Personnel Ltd          3000.00                  NaN   
3               Liquid Personnel Ltd          2791.67                  NaN   
4                         Eden Brown          2240.00                  NaN   
...                              ...              ...                  ...   
4995                    Flame Pharma          1916.67            Permanent   
4996  The Vacancy Management Company          1875.00  Fixed Term Contract   
4997                Live Recruitment          2708.33            Permanent   
4998                             NaN           417.17            Permanent   
4999                             NaN          1900.00  Fixed Term Contract   

                           Category              Closing          Location  \
0            Healthcare and Nursing  2013-02-22 15:00:00     Thames Ditton   
1            Healthcare and Nursing  2012-11-14 15:00:00        Shrewsbury   
2            Healthcare and Nursing  2012-12-08 00:00:00            London   
3            Healthcare and Nursing  2013-01-13 15:00:00           Bristol   
4            Healthcare and Nursing  2013-01-24 15:00:00             Wigan   
...                             ...                  ...               ...   
4995         Healthcare and Nursing  2013-10-24 15:00:00  Northern Ireland   
4996         Healthcare and Nursing  2012-06-20 12:00:00                UK   
4997  PR, Advertising and Marketing  2013-12-17 15:00:00     Hertfordshire   
4998                       Teaching  2012-03-15 12:00:00           Wrexham   
4999                       Teaching  2013-05-11 15:00:00         Salisbury   

     Full-Time Equivalent (FTE)              Opening  
0                     FULL TIME  2012-12-24 15:00:00  
1                     FULL TIME  2012-10-15 15:00:00  
2                     FULL TIME  2012-11-08 00:00:00  
3                     FULL TIME  2012-11-14 15:00:00  
4                     FULL TIME  2012-11-25 15:00:00  
...                         ...                  ...  
4995                        NaN  2013-09-24 15:00:00  
4996                        NaN  2012-05-21 12:00:00  
4997                  FULL TIME  2013-11-17 15:00:00  
4998                        NaN  2012-01-15 12:00:00  
4999                        NaN  2013-04-27 15:00:00  

[5000 rows x 10 columns]

#### Conflict 2: naming conflicts in the Title column and Job Title column

In [6]:
# Check the columns in df1
print(df1.columns)
# Check the columns in df2
print(df2.columns)

Index(['Id', 'Title', 'Location', 'Company', 'ContractType', 'ContractTime',
       'Category', 'Salary', 'OpenDate', 'CloseDate', 'SourceName'],
      dtype='object')
Index(['Id', 'Job Title', 'Organisation', 'Monthly Payment', 'Contract Type',
       'Category', 'Closing', 'Location', 'Full-Time Equivalent (FTE)',
       'Opening'],
      dtype='object')


The column `Title` in df1 and `Job Title` is the same thing, but their name is different. We just need to replace the name 'Job Title' in df2 with 'Title'.

In [7]:
# Rename the Job Title column 
df2 = df2.rename(columns = {'Job Title': 'Title'})
df2

Id                                              Title  \
0     95056457                Nursing Home Manager  Thames Ditton   
1     54116178      Deputy Manager  Nursing Home  RGN  Shrewsbury   
2     89491746         Deputy Manager  Nursing Home (RGN)  London   
3     31629963        Deputy Manager (RGN)  Nursing Home  Bristol   
4     77606265                                  RGN or RMN  Wigan   
...        ...                                                ...   
4995  63556254  Pharmaceutical  Analytical Chemist Jobs in Nor...   
4996  80341553     Medical Sales Representative  South East Essex   
4997  32782075  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
4998  64933538                     Curriculum Leader  Mathematics   
4999  83618722                TEACHER OF BUSINESS STUDIES AND LAW   

                        Organisation  Monthly Payment        Contract Type  \
0               Liquid Personnel Ltd          2750.00                  NaN   
1               Liquid Personnel Ltd          2583.33                  NaN   
2               Liquid Personnel Ltd          3000.00                  NaN   
3               Liquid Personnel Ltd          2791.67                  NaN   
4                         Eden Brown          2240.00                  NaN   
...                              ...              ...                  ...   
4995                    Flame Pharma          1916.67            Permanent   
4996  The Vacancy Management Company          1875.00  Fixed Term Contract   
4997                Live Recruitment          2708.33            Permanent   
4998                             NaN           417.17            Permanent   
4999                             NaN          1900.00  Fixed Term Contract   

                           Category              Closing          Location  \
0            Healthcare and Nursing  2013-02-22 15:00:00     Thames Ditton   
1            Healthcare and Nursing  2012-11-14 15:00:00        Shrewsbury   
2            Healthcare and Nursing  2012-12-08 00:00:00            London   
3            Healthcare and Nursing  2013-01-13 15:00:00           Bristol   
4            Healthcare and Nursing  2013-01-24 15:00:00             Wigan   
...                             ...                  ...               ...   
4995         Healthcare and Nursing  2013-10-24 15:00:00  Northern Ireland   
4996         Healthcare and Nursing  2012-06-20 12:00:00                UK   
4997  PR, Advertising and Marketing  2013-12-17 15:00:00     Hertfordshire   
4998                       Teaching  2012-03-15 12:00:00           Wrexham   
4999                       Teaching  2013-05-11 15:00:00         Salisbury   

     Full-Time Equivalent (FTE)              Opening  
0                     FULL TIME  2012-12-24 15:00:00  
1                     FULL TIME  2012-10-15 15:00:00  
2                     FULL TIME  2012-11-08 00:00:00  
3                     FULL TIME  2012-11-14 15:00:00  
4                     FULL TIME  2012-11-25 15:00:00  
...                         ...                  ...  
4995                        NaN  2013-09-24 15:00:00  
4996                        NaN  2012-05-21 12:00:00  
4997                  FULL TIME  2013-11-17 15:00:00  
4998                        NaN  2012-01-15 12:00:00  
4999                        NaN  2013-04-27 15:00:00  

[5000 rows x 10 columns]

#### Conflict 3: naming and value conflicts in the Company column and Organisation column

The `Company` column in df1 correspond to the `Organisation` column. We need to rename the `Organisation` column. Besides, there are missing values in the `Organisation` column. We need to replace them with 'non-specified'.

In [8]:
# Rename the Organisation column 
df2 = df2.rename(columns = {'Organisation': 'Company'})
# Replace the missing values with 'non-specified'
df2.loc[df2['Company'].isna(),'Company'] = 'non-specified'
df2

Id                                              Title  \
0     95056457                Nursing Home Manager  Thames Ditton   
1     54116178      Deputy Manager  Nursing Home  RGN  Shrewsbury   
2     89491746         Deputy Manager  Nursing Home (RGN)  London   
3     31629963        Deputy Manager (RGN)  Nursing Home  Bristol   
4     77606265                                  RGN or RMN  Wigan   
...        ...                                                ...   
4995  63556254  Pharmaceutical  Analytical Chemist Jobs in Nor...   
4996  80341553     Medical Sales Representative  South East Essex   
4997  32782075  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
4998  64933538                     Curriculum Leader  Mathematics   
4999  83618722                TEACHER OF BUSINESS STUDIES AND LAW   

                             Company  Monthly Payment        Contract Type  \
0               Liquid Personnel Ltd          2750.00                  NaN   
1               Liquid Personnel Ltd          2583.33                  NaN   
2               Liquid Personnel Ltd          3000.00                  NaN   
3               Liquid Personnel Ltd          2791.67                  NaN   
4                         Eden Brown          2240.00                  NaN   
...                              ...              ...                  ...   
4995                    Flame Pharma          1916.67            Permanent   
4996  The Vacancy Management Company          1875.00  Fixed Term Contract   
4997                Live Recruitment          2708.33            Permanent   
4998                   non-specified           417.17            Permanent   
4999                   non-specified          1900.00  Fixed Term Contract   

                           Category              Closing          Location  \
0            Healthcare and Nursing  2013-02-22 15:00:00     Thames Ditton   
1            Healthcare and Nursing  2012-11-14 15:00:00        Shrewsbury   
2            Healthcare and Nursing  2012-12-08 00:00:00            London   
3            Healthcare and Nursing  2013-01-13 15:00:00           Bristol   
4            Healthcare and Nursing  2013-01-24 15:00:00             Wigan   
...                             ...                  ...               ...   
4995         Healthcare and Nursing  2013-10-24 15:00:00  Northern Ireland   
4996         Healthcare and Nursing  2012-06-20 12:00:00                UK   
4997  PR, Advertising and Marketing  2013-12-17 15:00:00     Hertfordshire   
4998                       Teaching  2012-03-15 12:00:00           Wrexham   
4999                       Teaching  2013-05-11 15:00:00         Salisbury   

     Full-Time Equivalent (FTE)              Opening  
0                     FULL TIME  2012-12-24 15:00:00  
1                     FULL TIME  2012-10-15 15:00:00  
2                     FULL TIME  2012-11-08 00:00:00  
3                     FULL TIME  2012-11-14 15:00:00  
4                     FULL TIME  2012-11-25 15:00:00  
...                         ...                  ...  
4995                        NaN  2013-09-24 15:00:00  
4996                        NaN  2012-05-21 12:00:00  
4997                  FULL TIME  2013-11-17 15:00:00  
4998                        NaN  2012-01-15 12:00:00  
4999                        NaN  2013-04-27 15:00:00  

[5000 rows x 10 columns]

#### Conflict 4: naming and value conflicts in the Salary column and Monthly Payment column

The `Salary` column in df1 correspond to the `Monthly Payment` column. We need to rename the `Monthly Payment` column. Moreover, the value is annual salary in the `Salary` column but monthly salary in the `Monthly Payment` column, so we need to multiply the `Monthly Payment` column by 12. Also, we need to convert it with two decimal places.

In [9]:
# Rename the Organisation column 
df2 = df2.rename(columns = {'Monthly Payment': 'Salary'})
# Multiply the Salary column by 12
df2.loc[:,'Salary'] = df2['Salary'] * 12
# Convert the value to the value with two decimal places.
dfCopy = df2.copy() # To avoid SettingWithCopyWarning
dfCopy.loc[:,'Salary'] = df2['Salary'].apply(lambda x: format(x, '.2f'))
df2 = dfCopy
df2

Id                                              Title  \
0     95056457                Nursing Home Manager  Thames Ditton   
1     54116178      Deputy Manager  Nursing Home  RGN  Shrewsbury   
2     89491746         Deputy Manager  Nursing Home (RGN)  London   
3     31629963        Deputy Manager (RGN)  Nursing Home  Bristol   
4     77606265                                  RGN or RMN  Wigan   
...        ...                                                ...   
4995  63556254  Pharmaceutical  Analytical Chemist Jobs in Nor...   
4996  80341553     Medical Sales Representative  South East Essex   
4997  32782075  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
4998  64933538                     Curriculum Leader  Mathematics   
4999  83618722                TEACHER OF BUSINESS STUDIES AND LAW   

                             Company    Salary        Contract Type  \
0               Liquid Personnel Ltd  33000.00                  NaN   
1               Liquid Personnel Ltd  30999.96                  NaN   
2               Liquid Personnel Ltd  36000.00                  NaN   
3               Liquid Personnel Ltd  33500.04                  NaN   
4                         Eden Brown  26880.00                  NaN   
...                              ...       ...                  ...   
4995                    Flame Pharma  23000.04            Permanent   
4996  The Vacancy Management Company  22500.00  Fixed Term Contract   
4997                Live Recruitment  32499.96            Permanent   
4998                   non-specified   5006.04            Permanent   
4999                   non-specified  22800.00  Fixed Term Contract   

                           Category              Closing          Location  \
0            Healthcare and Nursing  2013-02-22 15:00:00     Thames Ditton   
1            Healthcare and Nursing  2012-11-14 15:00:00        Shrewsbury   
2            Healthcare and Nursing  2012-12-08 00:00:00            London   
3            Healthcare and Nursing  2013-01-13 15:00:00           Bristol   
4            Healthcare and Nursing  2013-01-24 15:00:00             Wigan   
...                             ...                  ...               ...   
4995         Healthcare and Nursing  2013-10-24 15:00:00  Northern Ireland   
4996         Healthcare and Nursing  2012-06-20 12:00:00                UK   
4997  PR, Advertising and Marketing  2013-12-17 15:00:00     Hertfordshire   
4998                       Teaching  2012-03-15 12:00:00           Wrexham   
4999                       Teaching  2013-05-11 15:00:00         Salisbury   

     Full-Time Equivalent (FTE)              Opening  
0                     FULL TIME  2012-12-24 15:00:00  
1                     FULL TIME  2012-10-15 15:00:00  
2                     FULL TIME  2012-11-08 00:00:00  
3                     FULL TIME  2012-11-14 15:00:00  
4                     FULL TIME  2012-11-25 15:00:00  
...                         ...                  ...  
4995                        NaN  2013-09-24 15:00:00  
4996                        NaN  2012-05-21 12:00:00  
4997                  FULL TIME  2013-11-17 15:00:00  
4998                        NaN  2012-01-15 12:00:00  
4999                        NaN  2013-04-27 15:00:00  

[5000 rows x 10 columns]

#### Conflict 5: naming and value conflicts in the ContractTime column and Contract Type column

In [10]:
# Check the values in the Contract Type column in df2
df2['Contract Type'].value_counts()

Permanent              3720
Fixed Term Contract     592
Name: Contract Type, dtype: int64

The `ContractTime` column in df1 correspond to the `Contract Type` column, so we need to rename the `Contract Type` column. Then, in the `Contract Type` column, we also need to change the values 'Permanent' to 'permanent', 'Fixed Term Contract' to 'contract' and missing values to 'non-specified', which will solve the value conflict.

In [11]:
# Rename the Contract Type column 
df2 = df2.rename(columns = {'Contract Type': 'ContractTime'})
# Change the values
df2.loc[df2['ContractTime'] == 'Permanent', 'ContractTime'] = 'permanent'
df2.loc[df2['ContractTime'] == 'Fixed Term Contract', 'ContractTime'] = 'contract'
df2.loc[df2['ContractTime'].isna(), 'ContractTime'] = 'non-specified'
df2['ContractTime'].value_counts()

permanent        3720
non-specified     688
contract          592
Name: ContractTime, dtype: int64

#### Conflict 6: value conflicts in the Category column

In [12]:
# Check the values in the Category column in df2
df2['Category'].value_counts()

Information Technology           1286
Engineering                      1066
Healthcare and Nursing            655
Finance and Accounting            637
Sales                             630
Hospitality and Catering          275
Teaching                          234
PR, Advertising and Marketing     217
Name: Category, dtype: int64

The values in the `Category` in df2 is similar to that in df1, we need to replace them with the values of the formats in df1 to fix the value conflict.

In [13]:
# Change the values
df2.loc[df2['Category'] == 'Information Technology', 'Category'] = 'IT Jobs'
df2.loc[df2['Category'] == 'Engineering', 'Category'] = 'Engineering Jobs'
df2.loc[df2['Category'] == 'Healthcare and Nursing', 'Category'] = 'Healthcare & Nursing Jobs'
df2.loc[df2['Category'] == 'Finance and Accounting', 'Category'] = 'Accounting & Finance Jobs'
df2.loc[df2['Category'] == 'Sales', 'Category'] = 'Sales Jobs'
df2.loc[df2['Category'] == 'Hospitality and Catering', 'Category'] = 'Hospitality & Catering Jobs'
df2.loc[df2['Category'] == 'Teaching', 'Category'] = 'Teaching Jobs'
df2.loc[df2['Category'] == 'PR, Advertising and Marketing', 'Category'] = 'PR, Advertising & Marketing Jobs'
df2['Category'].value_counts()

IT Jobs                             1286
Engineering Jobs                    1066
Healthcare & Nursing Jobs            655
Accounting & Finance Jobs            637
Sales Jobs                           630
Hospitality & Catering Jobs          275
Teaching Jobs                        234
PR, Advertising & Marketing Jobs     217
Name: Category, dtype: int64

#### Conflict 7: naming and value conflicts in the ContractTime column and Full-Time Equivalent (FTE) column

In [14]:
# Check the values in the Full-Time Equivalent (FTE) column in df2
df2['Full-Time Equivalent (FTE)'].value_counts()

FULL TIME    583
0.2 FTE       16
0.4 FTE       15
0.6 FTE       15
0.8 FTE       14
Name: Full-Time Equivalent (FTE), dtype: int64

The `ContractType` column in df1 correspond to the `Full-Time Equivalent (FTE)` column, so we need to rename the `Full-Time Equivalent (FTE)` column. Then, in the `Contract Type` column, we also need to change the values 'FULL TIME' to 'full_time’', '0.2 FTE', '0.4 FTE', '0.6 FTE', '0.8 FTE', to 'part_time' and missing values to 'non-specified', which will solve the value conflict.

In [15]:
# Rename the Full-Time Equivalent (FTE) column 
df2 = df2.rename(columns = {'Full-Time Equivalent (FTE)': 'ContractType'})
# Change the values
df2.loc[df2['ContractType'] == 'FULL TIME', 'ContractType'] = 'full_time'
df2.loc[df2['ContractType'] == '0.2 FTE', 'ContractType'] = 'part_time'
df2.loc[df2['ContractType'] == '0.4 FTE', 'ContractType'] = 'part_time'
df2.loc[df2['ContractType'] == '0.6 FTE', 'ContractType'] = 'part_time'
df2.loc[df2['ContractType'] == '0.8 FTE', 'ContractType'] = 'part_time'
df2.loc[df2['ContractType'].isna(), 'ContractType'] = 'non-specified'
df2['ContractType'].value_counts()

non-specified    4357
full_time         583
part_time          60
Name: ContractType, dtype: int64

#### Conflict 8: naming conflicts in the OpenDate column and Opening column, as well as CloseDate column and Closing

The `OpenDate` and `CloseDate` column in df1 correspond to the `Opening` and `Closing` column respectively, so we need to rename the two columns.

In [16]:
# Rename the two columns
df2 = df2.rename(columns = {'Opening': 'OpenDate', 'Closing': 'CloseDate'})
df2

Id                                              Title  \
0     95056457                Nursing Home Manager  Thames Ditton   
1     54116178      Deputy Manager  Nursing Home  RGN  Shrewsbury   
2     89491746         Deputy Manager  Nursing Home (RGN)  London   
3     31629963        Deputy Manager (RGN)  Nursing Home  Bristol   
4     77606265                                  RGN or RMN  Wigan   
...        ...                                                ...   
4995  63556254  Pharmaceutical  Analytical Chemist Jobs in Nor...   
4996  80341553     Medical Sales Representative  South East Essex   
4997  32782075  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
4998  64933538                     Curriculum Leader  Mathematics   
4999  83618722                TEACHER OF BUSINESS STUDIES AND LAW   

                             Company    Salary   ContractTime  \
0               Liquid Personnel Ltd  33000.00  non-specified   
1               Liquid Personnel Ltd  30999.96  non-specified   
2               Liquid Personnel Ltd  36000.00  non-specified   
3               Liquid Personnel Ltd  33500.04  non-specified   
4                         Eden Brown  26880.00  non-specified   
...                              ...       ...            ...   
4995                    Flame Pharma  23000.04      permanent   
4996  The Vacancy Management Company  22500.00       contract   
4997                Live Recruitment  32499.96      permanent   
4998                   non-specified   5006.04      permanent   
4999                   non-specified  22800.00       contract   

                              Category            CloseDate          Location  \
0            Healthcare & Nursing Jobs  2013-02-22 15:00:00     Thames Ditton   
1            Healthcare & Nursing Jobs  2012-11-14 15:00:00        Shrewsbury   
2            Healthcare & Nursing Jobs  2012-12-08 00:00:00            London   
3            Healthcare & Nursing Jobs  2013-01-13 15:00:00           Bristol   
4            Healthcare & Nursing Jobs  2013-01-24 15:00:00             Wigan   
...                                ...                  ...               ...   
4995         Healthcare & Nursing Jobs  2013-10-24 15:00:00  Northern Ireland   
4996         Healthcare & Nursing Jobs  2012-06-20 12:00:00                UK   
4997  PR, Advertising & Marketing Jobs  2013-12-17 15:00:00     Hertfordshire   
4998                     Teaching Jobs  2012-03-15 12:00:00           Wrexham   
4999                     Teaching Jobs  2013-05-11 15:00:00         Salisbury   

       ContractType             OpenDate  
0         full_time  2012-12-24 15:00:00  
1         full_time  2012-10-15 15:00:00  
2         full_time  2012-11-08 00:00:00  
3         full_time  2012-11-14 15:00:00  
4         full_time  2012-11-25 15:00:00  
...             ...                  ...  
4995  non-specified  2013-09-24 15:00:00  
4996  non-specified  2012-05-21 12:00:00  
4997      full_time  2013-11-17 15:00:00  
4998  non-specified  2012-01-15 12:00:00  
4999  non-specified  2013-04-27 15:00:00  

[5000 rows x 10 columns]

#### Conflict 9: missing conflicts in the SourceName column

There is no such column in df2 which correspond the SourceName column in df1, so we need to create a new column called SourceName in df2 and assign 'non-specified' to all values in the new column.

In [17]:
# Create a new column and assign 'non-specified' to all values in the new column
df2.loc[:,'SourceName'] = 'non-specified'
df2

Id                                              Title  \
0     95056457                Nursing Home Manager  Thames Ditton   
1     54116178      Deputy Manager  Nursing Home  RGN  Shrewsbury   
2     89491746         Deputy Manager  Nursing Home (RGN)  London   
3     31629963        Deputy Manager (RGN)  Nursing Home  Bristol   
4     77606265                                  RGN or RMN  Wigan   
...        ...                                                ...   
4995  63556254  Pharmaceutical  Analytical Chemist Jobs in Nor...   
4996  80341553     Medical Sales Representative  South East Essex   
4997  32782075  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
4998  64933538                     Curriculum Leader  Mathematics   
4999  83618722                TEACHER OF BUSINESS STUDIES AND LAW   

                             Company    Salary   ContractTime  \
0               Liquid Personnel Ltd  33000.00  non-specified   
1               Liquid Personnel Ltd  30999.96  non-specified   
2               Liquid Personnel Ltd  36000.00  non-specified   
3               Liquid Personnel Ltd  33500.04  non-specified   
4                         Eden Brown  26880.00  non-specified   
...                              ...       ...            ...   
4995                    Flame Pharma  23000.04      permanent   
4996  The Vacancy Management Company  22500.00       contract   
4997                Live Recruitment  32499.96      permanent   
4998                   non-specified   5006.04      permanent   
4999                   non-specified  22800.00       contract   

                              Category            CloseDate          Location  \
0            Healthcare & Nursing Jobs  2013-02-22 15:00:00     Thames Ditton   
1            Healthcare & Nursing Jobs  2012-11-14 15:00:00        Shrewsbury   
2            Healthcare & Nursing Jobs  2012-12-08 00:00:00            London   
3            Healthcare & Nursing Jobs  2013-01-13 15:00:00           Bristol   
4            Healthcare & Nursing Jobs  2013-01-24 15:00:00             Wigan   
...                                ...                  ...               ...   
4995         Healthcare & Nursing Jobs  2013-10-24 15:00:00  Northern Ireland   
4996         Healthcare & Nursing Jobs  2012-06-20 12:00:00                UK   
4997  PR, Advertising & Marketing Jobs  2013-12-17 15:00:00     Hertfordshire   
4998                     Teaching Jobs  2012-03-15 12:00:00           Wrexham   
4999                     Teaching Jobs  2013-05-11 15:00:00         Salisbury   

       ContractType             OpenDate     SourceName  
0         full_time  2012-12-24 15:00:00  non-specified  
1         full_time  2012-10-15 15:00:00  non-specified  
2         full_time  2012-11-08 00:00:00  non-specified  
3         full_time  2012-11-14 15:00:00  non-specified  
4         full_time  2012-11-25 15:00:00  non-specified  
...             ...                  ...            ...  
4995  non-specified  2013-09-24 15:00:00  non-specified  
4996  non-specified  2012-05-21 12:00:00  non-specified  
4997      full_time  2013-11-17 15:00:00  non-specified  
4998  non-specified  2012-01-15 12:00:00  non-specified  
4999  non-specified  2013-04-27 15:00:00  non-specified  

[5000 rows x 11 columns]

### 3. Merging data

The columns and values in df2 are now consistent with those in df1, so we can merge them into one dataframe.

In [18]:
# Merge two data sets
df = pd.concat([df1, df2])
df.index = range(0, df1.shape[0] + df2.shape[0]) # Assign new index to the new data set
df.loc[:,'Salary'] = df['Salary'].astype(float) # Make sure the value in Salary column is in the same format
df

Id                                              Title  \
0      71306331  Java Senior Developer  J****EE, JSF/GWT, Agile...   
1      69038357         Learning Technology Development Specialist   
2      71760708            Java Developer SQL  Mathematics / Logic   
3      69940759                         Web Designer  Bedfordshire   
4      72127424           Lead Systems Analyst,, Oracle Financials   
...         ...                                                ...   
55695  63556254  Pharmaceutical  Analytical Chemist Jobs in Nor...   
55696  80341553     Medical Sales Representative  South East Essex   
55697  32782075  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
55698  64933538                     Curriculum Leader  Mathematics   
55699  83618722                TEACHER OF BUSINESS STUDIES AND LAW   

               Location                         Company   ContractType  \
0                    UK                            VPSS  non-specified   
1                London           Randstad Technologies  non-specified   
2                London                   Client Server  non-specified   
3                    UK            Adapt IT Recruitment  non-specified   
4                Surrey          ARC IT Recruitment Ltd  non-specified   
...                 ...                             ...            ...   
55695  Northern Ireland                    Flame Pharma  non-specified   
55696                UK  The Vacancy Management Company  non-specified   
55697     Hertfordshire                Live Recruitment      full_time   
55698           Wrexham                   non-specified  non-specified   
55699         Salisbury                   non-specified  non-specified   

      ContractTime                          Category    Salary  \
0        permanent                           IT Jobs  42500.00   
1        permanent                           IT Jobs  34500.00   
2        permanent                           IT Jobs  47500.00   
3        permanent                           IT Jobs  21000.00   
4        permanent                           IT Jobs  80000.00   
...            ...                               ...       ...   
55695    permanent         Healthcare & Nursing Jobs  23000.04   
55696     contract         Healthcare & Nursing Jobs  22500.00   
55697    permanent  PR, Advertising & Marketing Jobs  32499.96   
55698    permanent                     Teaching Jobs   5006.04   
55699     contract                     Teaching Jobs  22800.00   

                  OpenDate            CloseDate     SourceName  
0      2013-01-23 15:00:00  2013-04-23 15:00:00   cwjobs.co.uk  
1      2013-04-27 15:00:00  2013-07-26 15:00:00   cwjobs.co.uk  
2      2012-07-09 15:00:00  2012-08-08 15:00:00   cwjobs.co.uk  
3      2012-03-07 00:00:00  2012-06-05 00:00:00   cwjobs.co.uk  
4      2012-12-29 00:00:00  2013-01-28 00:00:00   cwjobs.co.uk  
...                    ...                  ...            ...  
55695  2013-09-24 15:00:00  2013-10-24 15:00:00  non-specified  
55696  2012-05-21 12:00:00  2012-06-20 12:00:00  non-specified  
55697  2013-11-17 15:00:00  2013-12-17 15:00:00  non-specified  
55698  2012-01-15 12:00:00  2012-03-15 12:00:00  non-specified  
55699  2013-04-27 15:00:00  2013-05-11 15:00:00  non-specified  

[55700 rows x 11 columns]

### 4. Resolving data conflicts:

#### Conflict 1: complete duplications (e.g., complete duplications)

To check if there is any complete duplication, we should not contain the `Id` and `SourceName` column, because these two columns are the columns we created for merging two dataframes. We need to compare the remaining columns(the original columns in df2) for each row.

In [19]:
# Check if there is any complete duplication in terms of the columns except Id and SourceName
df.duplicated(subset = ['Title','Location','Company','ContractType','ContractTime','Category','Salary','OpenDate','CloseDate']).sum()

105

The result shows that there are 105 duplications in the new dataset. They are apparently redundant, so we need to delete them.

In [20]:
# Delete duplications
df = df[~df.duplicated(subset = ['Title','Location','Company','ContractType','ContractTime','Category','Salary','OpenDate','CloseDate'])]
df

Id                                              Title  \
0      71306331  Java Senior Developer  J****EE, JSF/GWT, Agile...   
1      69038357         Learning Technology Development Specialist   
2      71760708            Java Developer SQL  Mathematics / Logic   
3      69940759                         Web Designer  Bedfordshire   
4      72127424           Lead Systems Analyst,, Oracle Financials   
...         ...                                                ...   
55694  34143021  IT Infrastructure Specialist (Mlnlycke Health ...   
55695  63556254  Pharmaceutical  Analytical Chemist Jobs in Nor...   
55696  80341553     Medical Sales Representative  South East Essex   
55697  32782075  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
55698  64933538                     Curriculum Leader  Mathematics   

               Location                         Company   ContractType  \
0                    UK                            VPSS  non-specified   
1                London           Randstad Technologies  non-specified   
2                London                   Client Server  non-specified   
3                    UK            Adapt IT Recruitment  non-specified   
4                Surrey          ARC IT Recruitment Ltd  non-specified   
...                 ...                             ...            ...   
55694                UK                       Molnlycke  non-specified   
55695  Northern Ireland                    Flame Pharma  non-specified   
55696                UK  The Vacancy Management Company  non-specified   
55697     Hertfordshire                Live Recruitment      full_time   
55698           Wrexham                   non-specified  non-specified   

      ContractTime                          Category    Salary  \
0        permanent                           IT Jobs  42500.00   
1        permanent                           IT Jobs  34500.00   
2        permanent                           IT Jobs  47500.00   
3        permanent                           IT Jobs  21000.00   
4        permanent                           IT Jobs  80000.00   
...            ...                               ...       ...   
55694    permanent         Healthcare & Nursing Jobs  28500.00   
55695    permanent         Healthcare & Nursing Jobs  23000.04   
55696     contract         Healthcare & Nursing Jobs  22500.00   
55697    permanent  PR, Advertising & Marketing Jobs  32499.96   
55698    permanent                     Teaching Jobs   5006.04   

                  OpenDate            CloseDate     SourceName  
0      2013-01-23 15:00:00  2013-04-23 15:00:00   cwjobs.co.uk  
1      2013-04-27 15:00:00  2013-07-26 15:00:00   cwjobs.co.uk  
2      2012-07-09 15:00:00  2012-08-08 15:00:00   cwjobs.co.uk  
3      2012-03-07 00:00:00  2012-06-05 00:00:00   cwjobs.co.uk  
4      2012-12-29 00:00:00  2013-01-28 00:00:00   cwjobs.co.uk  
...                    ...                  ...            ...  
55694  2013-07-31 15:00:00  2013-10-29 15:00:00  non-specified  
55695  2013-09-24 15:00:00  2013-10-24 15:00:00  non-specified  
55696  2012-05-21 12:00:00  2012-06-20 12:00:00  non-specified  
55697  2013-11-17 15:00:00  2013-12-17 15:00:00  non-specified  
55698  2012-01-15 12:00:00  2012-03-15 12:00:00  non-specified  

[55595 rows x 11 columns]

#### Finding global key for the data
For the global key, the `Id` columns is the best choice, because the value of Id of each row is unique. We can check if they are unique in case something wrong happening.

In [21]:
# Code to identify global key
df.duplicated(subset = ['Id']).sum()

0

As you can see, each value in the `Id` column is unique. We choose the `Id` column as the global key for the dataset.

### 5. Saving the integrated and reshaped data
The last part of the integration process is to export our output data to csv format, named as:
- 'S3749857_dataset_integrated.csv'

In [22]:
# Save output data
df.to_csv('S3749857_dataset_integrated.csv', index = False)

## Summary of the Assessment Task
This assessment task is very comprehensive for cleaning data. I have learnt a lot in this assessment. 